lib/mqtt/MqttController.js

```diff
@@ -0,0 +1,595 @@
+     *
+     * @param {import("./handles/MqttHandle")} handle
+     * @return {Promise<void>}
+     */
+    async dropHomieAttributes(handle) {
+        if (!this.homieEnabled) {
+            return;
+        }
+        if (this.isInitialized()) {
+            throw new Error("Homie attributes may be altered only while the MQTT controller is not initialized");
+        }
+        const attrs = handle.getHomieAttributes();
+        const baseTopic = handle.getBaseTopic();
+
+        for (const topic of Object.keys(attrs)) {
+            // @ts-ignore
+            await this.asyncClient.publish(baseTopic + "/" + topic, "", {
+                // @ts-ignore
+                qos: MqttCommonAttributes.QOS.AT_LEAST_ONCE,
+                retain: false
+            });
+        }
+    }
+
+    /**
+     * @callback withHassAsyncCb
+     * @param {import("./homeassistant/HassController")} hass
+     * @return {Promise<void>}
+     */
+    /**
+     * This is a fancy wrapper inspired by Python's "with" statement. It should be called by Handles that intend to
+     * register their own Hass components. The callback will only be called if Hass is actually enabled.
+     *
+     * @param {withHassAsyncCb} callback
+     * @return {Promise<void>}
+     */
+    async withHassAsync(callback) {
+        if (this.hassEnabled) {
+            await callback(this.hassController);
+        }
+    }
+
+
+    /**
+     * @callback withHassCb
+     * @param {import("./homeassistant/HassController")} hass
+     * @return {void}
+     */
+    /**
+     * This is a fancy wrapper inspired by Python's "with" statement. It should be called by Handles that intend to
+     * register their own Hass components. The callback will only be called if Hass is actually enabled.
+     *
+     * @param {withHassCb} callback
+     * @return {void}
+     */
+    withHass(callback) {
+        if (this.hassEnabled) {
+            callback(this.hassController);
+        }
+    }
+
+
+    /**
+     * This is effectively a wrapper around mqtt.publish. However, this is intended for *exclusive* usage by
+     * HassController in order to enforce responsibility separation.
+     * No class that doesn't end in "Controller" is allowed to publish directly.
+     * All classes not ending in "Controller" must ask their respective controller to fetch their data and publish it.
+     *
+     * If you find yourself wanting to use this outside of HassController, please start a discussion on GitHub and
+     * tag @Depau.
+     *
+     * @package
+     * @param {string} topic
+     * @param {string} message
+     * @param {object} [options]
+     * @return {Promise<void>}
+     */
+    async publishHass(topic, message, options) {
+        const err = new Error();
+        Error.captureStackTrace(err, this.publishHass);
+        const lines = err.stack.split("\n");
+        if (lines !== undefined && lines.length > 1 && !lines[1].includes(HassController.name)) {
```

Comment:
Lol this level of self-defense

We should keep in mind that creating a stacktrace is a somewhat expensive operation, although that shouldn't really matter here, because it will only be called every few seconds, I think?

Reply:
I just have this feeling that someone at some point is gonna use it to duct-tape some weird behavior for some "special" hass component and that it will slip through review.

But maybe it's just:

> +++ BREAKING NEWS: "Man invents fictional scenario and gets angry about it" +++

I guess it can be dropped :)